lib/saxes.js

```diff
@@ -488,13 +487,20 @@ class SaxesParser {
    * check for issues that are outside the scope of this project and can report
    * errors.
    *
-   * @param {Error} er The error to report.
+   * @param {String} er The error to report.
    *
    * @returns this
    */
   fail(er) {
-    const message = (this.trackPosition) ?
-          `${this.fileName}:${this.line}:${this.column}: ${er}` : er;
+    const fragments = [];
+    if (this.fileName) {
+      fragments.push(this.fileName);
+    }
+    if (this.trackPosition) {
+      fragments.push(this.line, this.column);
+    }
+    const message = fragments.length > 0 ?
+      `${fragments.join(":")}: ${er}` : er;
```

Comment:
The approach here is not optimal. There are languages in which if you are building a string from multiple smaller strings it is better to plunk the strings into an array and then join them as you do here. For instance when string concatenation is implemented naively: allocate a new string object which is at least as long as the two strings to be concatenated, copy the first string at the start of the new string, copy the 2nd string after the first, return the new string. Such naive implementations are inefficient because there's a lot of memory churn and parts of the final string are copied over and over. (e.g. if you concatenate 10 strings together, the bytes of the 1st string are going to be copied 9 times before you get the final string.)

However, v8, which is used in Chrome and Node, does not do string concatenations naively. v8 uses multiple implementations of the `string` primitive. When you concatenate strings, it essentially creates a linked list of string fragments. So it does not spend its time allocating and copying over and over. (I don't quite know how Firefox does it, but it uses something having a similar effect as what v8 does.)

Here's an example of the performance difference: 

https://jsperf.com/error-message-with-push-join-or-string-concatenation

Doing string concatenation runs twice as fast on my machine.

Reply:
Makes sense! Fixed in 1ebdfe52e6827ede3e37c991f33ca3592a80fbb8